### ISSUE TITLE
Reasonable people are complaining about Agda's unreasonable packaging

### ISSUE THREAD

**Issue by JacquesCarette** [agda/agda#5725]
![image](https://user-images.githubusercontent.com/1855141/148481601-d8f39c0f-9160-4a44-b4ff-b875a448b0fd.png)

This is rather embarrassing, no?  Shouldn't a "standard install" of Agda 1) include the library, and 2) be 'simple' (i.e. not involve editing any text files)?  Wouldn't it be even better if it didn't take 30-40 minutes?

Assuming that the answer to the above is "yes, such a standard install should exist", I'd be happy to coordinate the effort.

**Comment by gallais** [comment]
Annoyances aside, a lot of newcomers seem to also be tripped over by the fact
you have to [pick a specific version of the standard library](https://wiki.portal.chalmers.se/agda/Libraries/StandardLibrary) that is compatible with
the version of Agda you are installing.

As for the solution, I don't know if it's an installer or rather a package manager
that we need the most.

**Comment by jespercockx** [comment]
I very much applaud this initiative! Yes, there should absolutely be an easy way to get Agda installed and running with as little friction as possible. However, there are the usual questions:

1. How do you make sure that the installer is up-to-date with the latest release of Agda and the standard library?
2. How do you support upgrading and/or uninstalling?
3. Do you also package a compatible version of GHC so people can compile their code? If yes, how do you make sure this does not conflict/override an existing installation of GHC?
4. Do you also package an IDE (Emacs / VS Code / ...) with agda-mode installed? If yes, how do you make sure this does not conflict with/override an existing installation of this IDE?
5. Do you have a separate installer for each supported platform (Linux/MacOS/Windows)? Or do you expect Windows users to use WSL? If Windows is supported, where do you put the `.agda` folder with all the configurations for the libraries?
6. How about different Linux distros, do you just have a `.deb`/`.rpm` or do you use something more universal such as FlatPak/AppImage/Snap/...?
7. How do you ensure the installer includes all the licenses for the different open-source libraries that Agda relies on?

**Comment by nad** [comment]
Can we learn from other projects?

I recently installed Coq 8.13.2:
* The package available via my Linux distribution's package manager was not sufficiently recent.
* One can also use a "snap", but not for the 32-bit system that I was using.
* I installed opam, some GTK libraries, as well as (parts of) the Coq platform. This involved running some shell-script, which installed programs in a non-standard location (`~/.opam`). The script also edited `~/.bash_profile` (and I don't want scripts to mess with my configuration files).

I also recently installed Agda on a new machine. I guess this involved using the package manager to install GHC, Alex, Happy and Cabal, and then compiling Agda using `make install` (this can involve installation of files in a non-standard location, `~/.cabal`, and the user's Emacs configuration might be changed). I copied the files in `~/.agda` from a different machine, but I guess I had to edit something, because I did not install all the libraries that I had installed on the other machine.

Back in 2018 I installed Lean 3.4.1. The binary I found did not work on my 32-bit system, but my installation log suggests that compilation from scratch was straightforward (once some dependencies had been installed using the package manager), and I could choose where the files were installed. A library seems to have been included.

Last year I installed Idris 2. My installation log suggests that compilation from scratch was straightforward (once some dependencies had been installed using the package manager), and I could choose where the files were installed.

I have also installed Matita, using my system's package manager. However, Matita's last release seems to be from 2016.

**Comment by xekoukou** [comment]
I personally use nix. It installs all the dependencies without affecting the system.

Every few weeks, I upgrade agda from github. If new ghc and other package versions are needed, it automatically performs the upgrade to those packages.
(It does that by readying the cabal file.)

It works on all linux distros, MacOS and WIndows. At least that's what it says in the site.

Maybe @shlevy could help in this. He has already contributed a Flake.nix file in this repository.

**Comment by shlevy** [comment]
I'd be happy to help with the actual implementation if we use Nix, but I'd prefer to defer to more serious Agda users as to the right bundling of various components to expose to users.

**Comment by jespercockx** [comment]
The target audience here are users who want things to "just work" without having to configure a ton of things by hand. I am not a Nix user myself so I might be wrong, but my impression is that it is rather unlikely that this target audience will be satisfied with a solution that requires them to use Nix. Unless perhaps there is a way to abstract away all the details and create some kind of installer that uses Nix under the hood?

**Comment by shlevy** [comment]
Yes, Nix itself is a challenge if you don't already have it set up. There are definitely ways to hide the Nix usage (either producing truly Nix-independent artifacts or hiding the Nix store usage through startup namespace trickery) but I don't know if that will be the best fit for your needs.

**Comment by JacquesCarette** [comment]
I agree with @jespercockx that the main target audience here should be users who want things to "just work". So that calls for some kind of installer?

As to the list of issues brought up earlier (also by @jespercockx ):
- the installer should be opinionated: it makes choices of what it does and doesn't install. It should be optimized for "just work".
- versioning issues: see above. A release of an installer would be an explicit bundle of choices, that ought to "just work". In particular, it would pick a version of the tool-chain that is known to work.
- how many should be offered? As many as the open source volunteers are willing to develop.
- regarding packaging an IDE: my default would be a resounding "Yes!", BUT if it's clear that doing so is quite likely to muck up most users' configs, then I would drop that. Justification: "just work". Things don't 'just work' if they #$%#$ up a user's environment.
- (lots of technical questions skipped. yes, these choices all have to be made. this is not easy.)

As to @nad 's exhortation to learn from other projects: we can only learn from those that are 'successful'. I tried to install Idris 2 (on my Windows box, shudder) and failed. I should re-try on my linux laptop.  But the larger point: most of us involved in this thread are not part of the target audience. We know too much and we're willing to go through more steps than most. Plus we tend to want bleeding edge, not "bog standard".

But still, let's dig deeper.  Let's google "install X":
- [Coq](https://coq.inria.fr/download). Provides binary installers for Windows and Mac.  The "Coq Platform" provides various levels of built-in libraries.
- [Lean](https://leanprover-community.github.io/get_started.html). Comes with a library. The regular install is extremely complicated, but on offer there are also other ways to get started that "just work". They are just opinionated about users should do because they're biased towards getting people to contribute.
- [Idris 2](https://www.idris-lang.org/pages/download.html). Comes with a library. Quite hard to install. Oddly, there's an alternate [install page](https://github.com/idris-lang/Idris2/blob/main/INSTALL.md) that mentions simpler ways to do it.
- [Isabelle](https://isabelle.in.tum.de/installation.html). Provides either Docker image with everything, or installers with everything.

It's important to remember that a lot of users of Agda will compare the experience to a lot of other software they've installed, such as Chrome, Firefox, Thunderbird, Zoom, Discord, Zulip, etc (just to quickly look at my own desktop). Yes, it's a completely unfair comparison!

**Comment by L-TChen** [comment]
By the way, we still need to resolve the license issue of distributing binaries. #5202

**Comment by jpoiret** [comment]
Hi, I'm popping here to say that we could probably provide an environment variable like `C_INCLUDE_PATH` that would point to a folder containing Agda libraries, that could get automatically loaded without needing to touch `.agda/libraries`. I have an experimental patch for this that I used to test easier packaging of libraries in Guix (a Nix derivative): you would only need to install `agda` and `agda-stdlib` and it would just work out of the box.

For an easier installation path that doesn't require being a Nix or Guix user, both tools have `nix bundle` or `guix pack` to create standalone archives that should work on any machine running Linux. I just tried it out on guix with `guix pack -RR -S /bin=bin /etc=etc emacs emacs-agda2-mode agda` to create the relocatable pack, then you can just unpack that anywhere, run `source etc/profile` in a shell then `emacs`, you can load an agda file without messing with any init.el file and it will Just Work™ (I tried this in a container with just coreutils and bash available). This uses linux user namespaces if they're available, otherwise it falls back to libfakeroot or other execution engines. With a mechanism such as the one I outlined above, you could even add libraries to that pack and they would be immediately accessible (you would still need to add them to your `agda-lib` or `defaults` though).

Guix doesn't support MacOS but Nix does (but no `nix bundle` support there), and regarding Windows the easiest solution is probably to use WSL instead of messing about with actual native solutions.

The n°1 thing I want Agda to avoid is to re-invent the wheel with packaging, there's no need for the project to spend time on a new tool that is Agda-specific when other tools already exist to do the exact same thing.
